panels/metrics/cc-metrics-panel.c

```diff
@@ -190,25 +179,29 @@ cc_metrics_panel_constructed (GObject *object)
                         G_CALLBACK (on_metrics_proxy_properties_changed), self);
 
       value = g_dbus_proxy_get_cached_property (self->metrics_proxy, "Enabled");
-      self->metrics_active = g_variant_get_boolean (value);
+      metrics_active = g_variant_get_boolean (value);
       g_clear_pointer (&value, g_variant_unref);
     }
 
   permission = polkit_permission_new_sync ("com.endlessm.Metrics.SetEnabled",
                                            NULL, NULL, NULL);
-  if (!permission)
-    metrics_can_change = FALSE;
-  else
+  if (permission)
     metrics_can_change = g_permission_get_allowed (permission);
 
   box = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 6);
-  gtk_widget_show (box);
+  gtk_widget_set_visible (box, true);
 
   gtk_widget_set_sensitive (GTK_WIDGET (self->enable_metrics_switch), metrics_can_change);
-  gtk_switch_set_state (GTK_SWITCH (self->enable_metrics_switch), self->metrics_active);
+
+  /* It would be clearer to use our own metrics_enabled property, but for this
+   * purpose it is fine to treat the switch's state property as canonical. */
+  g_object_bind_property (self->enable_metrics_switch, "state",
+                          self->enable_metrics_switch, "active",
+                          G_BINDING_SYNC_CREATE);
```

Comment:
I don't quite understand why this is needed. Is it to handle the daemon property changing behind this page's back?

I looked at how another page in this app, the Thunderbolt one in particular, handles this. There, on initialisation they block the `state-set` handler, then set the `active` property (which causes the default `state-set` handler to run which just sets `state`).

https://gitlab.gnome.org/GNOME/gnome-control-center/-/blob/3b22ae0da6aed94c8226f03194aee64a159d2653/panels/privacy/cc-bolt-page.c#L467-471

Reply:
Ooh, I see how they did that. That's much nicer, indeed. I switched over to that pattern in the newest commit. Unfortunately, in my testing, the default state-set handler isn't running / isn't setting the "active" property, which is strange and confusing but I unfortunately can't dig into it very well because of a sad combination of slow internet and oddly sluggish laptop. I worked around it by calling both `gtk_switch_set_state ()` and `gtk_switch_set_active ()` in a new `cc_metrics_panel_enable_sync` function, which is lame but at least contains the weird hacky stuff to one place. If it bothers you, I can take a better look on Wednesday when I'm back in Canada with a big computer.